### PR TITLE
Fix race conditions in CREATE/DROP of different replicas of ReplicatedMergeTree

### LINKED ISSUE #11432: Exception: Directory for table data already exists.
Version: `20.4.4.18`

If table creation failed, second `create` query may throw exception `Directory for table data ... already exists.`

Examples:
```
CREATE TABLE ...

Code: 497. DB::Exception: Received from ... DB::Exception: travel-test: Not enough privileges. ...

CREATE TABLE ...
Received exception from server (version 20.4.4):
Code: 57. DB::Exception: Received from localhost:9000. DB::Exception: Directory for table data ... already exists. 
```

```
CREATE TABLE ...

Received exception from server (version 20.4.4):
Code: 225. DB::Exception: Received from localhost:9000. DB::Exception: Can't create replicated table without ZooKeeper. 


CREATE TABLE ...
Received exception from server (version 20.4.4):
Code: 57. DB::Exception: Received from localhost:9000. DB::Exception: Directory for table data ... already exists. 
```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix race conditions in CREATE/DROP of different replicas of ReplicatedMergeTree. Continue to work if the table was not removed completely from ZooKeeper or not created successfully. This fixes #11432.

Detailed description / Documentation draft:
See https://clickhouse-test-reports.s3.yandex.net/11580/e2c6d090192d870edffc40e40deeb4c14417e5be/functional_stateless_tests_(thread)/test_run.txt.out.log

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
PVS Ok.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -403,46 +417,306 @@ void StorageReplicatedMergeTree::createNewZooKeeperNodes()
...
+        LOG_DEBUG(log, "Creating table {}", zookeeper_path);
+
+        /// We write metadata of table so that the replicas can check table parameters with them.
+        String metadata = ReplicatedMergeTreeTableMetadata(*this).toString();
+
+        Coordination::Requests ops;
+        ops.emplace_back(zkutil::makeCreateRequest(zookeeper_path, "", zkutil::CreateMode::Persistent));
+
+        /// Check that the table is not being dropped right now.
+        ops.emplace_back(zkutil::makeCreateRequest(zookeeper_path + "/dropped", "", zkutil::CreateMode::Persistent));
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:491]: interesting way to check that node doesn't exist)

**Review by alesapin — APPROVED** [review]
I think it's quite a complicated solution, but I cannot propose something simpler. It will not work between concurrent drop/create between different versions but it's OK.

**Review by tavplubix — APPROVED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -403,46 +417,306 @@ void StorageReplicatedMergeTree::createNewZooKeeperNodes()
...
+                code = zookeeper->tryMulti(ops, responses);
+
+                if (code == Coordination::ZNONODE)
+                {
+                    LOG_WARNING(log, "Table {} is already finished removing by another replica right now", replica_path);
+                }
+                else if (code == Coordination::ZNOTEMPTY)
+                {
+                    throw Exception(fmt::format(
+                        "The old table was not completely removed from ZooKeeper, {} still exists and may contain some garbage.", zookeeper_path), ErrorCodes::TABLE_WAS_NOT_DROPPED);
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Storages/StorageReplicatedMergeTree.cpp:468]: I think that it's not possible. And it does not reproduce in test.
I will change to code to LOGICAL_ERROR.

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply src/Storages/StorageReplicatedMergeTree.cpp:491]: Yes, and I did not invent it :) It's already used in other places in code.

**Pull request merged**